File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        int[] chars = new int[26];
        for(char c : p.toCharArray()) chars[c-'a']++;

        int left = 0;
        int right = 0;
        int cnt = p.length();

        while(right < s.length()){
            if(chars[s.charAt(right++) - 'a']-- >= 1) cnt--;
            if(cnt == 0) result.add(left);

            if(right - left == p.length() && chars[s.charAt(left++) - 'a']++ >= 0) cnt++;
        }
        return result;
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> res;
        map<char, int> m;
        for(char c = 'a'; c < 26; c++) m[c] = 0;

        for(int i = 0; i < p.size(); i++) m[p[i]]++;

        int left = 0, right = 0, cnt = p.size();

        while(right < s.size()) {
            if(m[s[right]] >= 1) cnt--;
            m[s[right]]--;
            right++;
            if(cnt == 0) res.push_back(left);

            if(right - left == p.size()) {
                if(m[s[left]] >= 0) cnt++;
                m[s[left]]++;
                left++;
            }
        }
    return res;
    }
};